README.rst
django-admin-bootstrapped
=========================

A Django admin theme using Bootstrap. It doesn't need any kind
of modification on your side, just add it to the installed apps.

Requirements
------------

-  Django ``>=2.0,<3.0``

Installation
------------

1. Download it from PyPi with ``pip install django-admin-bootstrapped``
2. Add into the ``INSTALLED_APPS`` **before** ``'django.contrib.admin'``:

::

    'django_admin_bootstrapped',

3. Have fun!

Configuration
-------------

For a full bootstrap3 experience you may want to use a custom renderer for the fields.
There's one available in tree that requires the ``django-bootstrap3`` application installed.
You have to add to your project settings file:
::

    DAB_FIELD_RENDERER = 'django_admin_bootstrapped.renderers.BootstrapFieldRenderer'


`Messages <http://docs.djangoproject.com/en/dev/ref/contrib/messages>`__ will have ``alert-info`` tag by default,
so you may want to add Bootstrap 3 tags for different message levels to make them styled appropriately.
Add to your project settings file:
::

    from django.contrib import messages

    MESSAGE_TAGS = {
                messages.SUCCESS: 'alert-success success',
                messages.WARNING: 'alert-warning warning',
                messages.ERROR: 'alert-danger error'
    }

Now, adding messages like this:
::

    messages.success(request, "My success message")
    messages.warning(request, "My warning message")
    messages.error(request, "My error message")

will result into this:

.. image:: https://i.imgur.com/SQNMZZE.png

Goodies
-------

Add custom html to the change form of any model with a template
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can inject custom html on top of any change form creating a template
named ``admin_model_MODELNAME_change_form.html`` into the application's
template folder. Eg:
``myapp/templates/myapp/admin_model_mymodelname_change_form.html`` or
``project/templates/myapp/admin_model_mymodelname_change_form.html``.

Inline sortable
~~~~~~~~~~~~~~~

You can add drag&drop sorting capability to any inline with a couple of
changes to your code.

First, add a ``position`` field in your model (and sort your model
accordingly), for example:

::

    class TestSortable(models.Model):
        that = models.ForeignKey(TestMe)
        position = models.PositiveSmallIntegerField("Position")
        test_char = models.CharField(max_length=5)

        class Meta:
            ordering = ('position', )

Then in your admin.py create a class to handle the inline using the
``django_admin_bootstrapped.admin.models.SortableInline`` mixin, like
this:

::

    from django_admin_bootstrapped.admin.models import SortableInline
    from models import TestSortable

    class TestSortable(admin.StackedInline, SortableInline):
        model = TestSortable
        extra = 0

You can now use the inline as usual. See the screenshots section to see what the result
will look like.

This feature was brought to you by `Kyle Bock <https://github.com/kwbock>`__. Thank you Kyle!


XHTML Compatible
~~~~~~~~~~~~~~~~

Compatible with both html and xhtml. To enable xhtml for your django app
add the following to your settings.py: DEFAULT\_CONTENT\_TYPE =
'application/xhtml+xml'

Generic lookups in admin
~~~~~~~~~~~~~~~~~~~~~~~~

All that needs to be done is change the admin widget with either
formfield\_overrides like this:

::

    from django_admin_bootstrapped.widgets import GenericContentTypeSelect

    class SomeModelAdmin(admin.ModelAdmin):
        formfield_overrides = {
            models.ForeignKey: {'widget': GenericContentTypeSelect},
        }

Or if you want to be more specific:

::

    from django_admin_bootstrapped.widgets import GenericContentTypeSelect

    class SomeModelAdmin(admin.ModelAdmin):
        def formfield_for_dbfield(self, db_field, **kwargs):
            if db_field.name == 'content_type':
                kwargs['widget'] = GenericContentTypeSelect
            return super(SomeModelAdmin, self).formfield_for_dbfield(db_field, **kwargs)

If you decide on using ``formfield_overrides`` `you should be aware of
its limitations with relation
fields <https://docs.djangoproject.com/en/dev/ref/contrib/admin/#django.contrib.admin.ModelAdmin.formfield_overrides>`__.

This feature (and many more) was brought to you by `Jacob
Magnusson <https://github.com/jmagnusson>`__. Thank you Jacob!

Contributing
------------

Every code, documentation and UX contribution is welcome.

Found an issue? Report it in the bugtracker!

Have some free time? Help fixing an already filed issue, just remember to work on a separate branch please.

Screenshots
-----------

Homepage
~~~~~~~~

.. image:: https://cloud.githubusercontent.com/assets/12932/6967318/d7064abe-d95e-11e4-91bc-6de527550557.png

List view with filters in dropdown
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: https://cloud.githubusercontent.com/assets/12932/6967319/d71a9c6c-d95e-11e4-86cf-47e8857582c1.png

Change form view
~~~~~~~~~~~~~~~~

.. image:: https://cloud.githubusercontent.com/assets/12932/6966950/98661ba6-d95c-11e4-8bb3-e4b18759115b.png

.. |PyPI version| image:: https://pypip.in/d/django-admin-bootstrapped/badge.png
   :target: https://pypi.python.org/pypi/django-admin-bootstrapped


CHANGELOG
==== 2.7.0 (2019-10-31) ====

- Use inline has_change_permission to fix display of last item (Augusto Men)

==== 2.6.2 (2019-07-10) ====

- Add Django >=2.0,<3.0 support (Dominik Magdaleński)

==== 2.6.1 (2018-05-23) ====

- Fix rendering of "related" widgets (related_widget_wrapper.html template) (Dominik Magdaleński)

==== 2.6.0 (2018-02-01) ====

- Drop support of Django 1.8 (Dominik Magdaleński)
- Add support of Django 2.0 (Dominik Magdaleński)

==== 2.5.7 (2015-12-15) ====

- travis-ci integration (Jacek Wielemborek)
- Handle fieldsets with empty lines (Kevin Gu)
- Require django < 1.9 (Riccardo)

==== 2.5.6 (2015-09-27) ====

- Fix data-admin-url model lookup for Django 1.8 ContentType #218 (Andrew Yager)
- Fix glyphicon display for generic-lookup #214 (Andrew Yager)
- Add space before caret in language selector #219 (Benoit Formet)
- Fix duplicate error messages in tabular inlines #220 (Josh Kelley)

==== 2.5.5 (2015-09-05) ====

- Fix current page highlightning #207 (Riccardo)
- Cleanup style in templatetags (Bang Dao)
- Bootstrapify non-form and non-field errors for inlines #213 (Andrei Menkou, Riccardo)

==== 2.5.4 (2015-07-02) ====

- Boostrap more auth/user/change_password.html (Josh Kelley)
- Cleanup filter-horizontal style (Walker Boyle, Riccardo)
- Cleanup file-upload widget (Walker Boyle)
- Bootstrapify admin/related_widget_wrapper.html (Riccardo)

==== 2.5.2 (2015-06-24) ====

- Fix user model username handling #200 (Riccardo)
- Remove duplicated breadcrumbs dividers #201 (Riccardo)
- Bootstrapify auth/user/change_password.html (Riccardo)
- Upgrade bootstrap to 3.3.5 #202 (Riccardo)

==== 2.5.1 (2015-06-15) ====

- Cleanup time and m2m widget rendering (Riccardo)
- Fix AdminSplitDateTime bootstrap3 rendering #168 (Riccardo)
- Support setting admin site title as upstream django #195 (Riccardo)

==== 2.5.0 (2015-04-24) ====

- Bump minimum django version to 1.8 (Riccardo)
- Hide custom way to override admin and app titles (Riccardo)
- Update templates to match django 1.8 features (Riccardo)
- Convert more fields to dab_field_rendering (Riccardo)
- Add RTL support (Ali Vakilzade)
- Add missing locale middleware to test project (Riccardo)

==== 2.4.0 (2015-04-02) ====

WARNING: we plan to do a very quick cycle for this series and move to django 1.8 support only

- bootstrap3 sub-application has been removed (Riccardo)
- Drop bootstrap3 theme for a flat look (Riccardo)
- Drop unmaintaned django-cms integration #52, #112 (Riccardo)
- Cleanup change list actions spacing #186 (David Rideout)
- Bump minimum django version to 1.7 (Riccardo)
- Hide fieldset rows without visible fields (Sylvain Fankhauser)
- Update screenshots in README.rst #187 (Riccardo)
- Update bootstrap to 3.3.4 #184 (Riccardo)

==== 2.3.5 (2015-03-13) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Reverted commit that broken fieldsets on django 1.6. Oops! (Riccardo)

==== 2.3.4 (2015-03-10) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Django 1.8 compatibility fixes #177, #178 (Joel Burton)
- Fixed some glyphicons classes conversion from bootstrap 2 (Riccardo)
- Django 1.9 compatibility fixes (Riccardo)
- Fix running test project under a wsgi server (Riccardo)
- Cleanup change list actions and pagination a bit (Riccardo)
- Hide fieldset rows without visible fields (Sylvain Fankhauser)
- Add ability to run tests from setup (Riccardo)

==== 2.3.3 (2015-02-10) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Import firstof from future to ensure Django 1.8 compatibility #175 (Jamie Curle)
- Fix CheckboxSelectMultiple style #174 (Riccardo)
- Fix input-group-addon rendering #172 (Riccardo)

==== 2.3.2 (2015-01-23) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Cleanup messages tags usage in alerts #166 (Patrick Miller)
- Cope with django-bootstrap3 api update #171 (Fernando Gonzalez)
- Update bootstrap to 3.3.2 #170 (Riccardo)
- Make the filter menu in change list scrollable #169 (Riccardo)

==== 2.3.1 (2014-12-08) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Add missing import in custom field rendering #162 (James Oakley)
- Add missing timezone handling to base template #164 (qxcv)
- Add custom field rendering tests #163 (Riccardo)

==== 2.3.0 (2014-12-04) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Fix "Add model" button focus #156 (Riccardo)
- Cleanup django-filer templates (UI is terrible though) #158 (Riccardo)
- Add support for custom field rendering #152 (Riccardo)

==== 2.2.1 (2014-11-26) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Fix "Add model" button hover #156 (Riccardo)
- Remove spurious hamburger icon in index (Riccardo)
- Fix collapsible handling with stacked inlines #129 (Brandon Cazander)

==== 2.2.0 (2014-11-19) ====

WARNING: bootstrap3 application is deprecated, would be removed in the next major release
Please remove it from the INSTALLED_APPS in settings.py

- Move bootstrap3 static and templates back to django admin bootstrapped (Riccardo)
- Upgrade bootstrap to 3.3.1 (Riccardo)
- Make inner header responsive #144 (Riccardo)

==== 2.1.1 (2014-11-10) ====

- Add missing translation to search button (Chun-Wei Shen)
- Fix js error on add view #155 (Riccardo, Marc)
- Fix opts.module_name RemovedInDjango18Warning #154 (Riccardo)


test_django_admin_bootstrapped/test_django_admin_bootstrapped/models.py
from django.db import models
from django.core.validators import validate_comma_separated_integer_list


class TestMe(models.Model):
    test_m2m = models.ManyToManyField('self', blank=True, help_text="Lorem dolor")
    test_ip = models.GenericIPAddressField(help_text="Lorem dolor")
    test_url = models.URLField(help_text="Lorem dolor")
    test_int = models.IntegerField(help_text="Lorem dolor")
    test_img = models.ImageField(upload_to='dummy', blank=True)
    test_file = models.FileField(upload_to='dummy', blank=True)
    test_date = models.DateField(help_text="Lorem dolor")
    test_char = models.CharField(max_length=50, help_text="Lorem dolor")
    test_bool = models.BooleanField(help_text="Lorem dolor", default=False)
    test_time = models.TimeField(help_text="Lorem dolor")
    test_slug = models.SlugField(help_text="Lorem dolor")
    test_text = models.TextField(help_text="Lorem dolor")
    test_email = models.EmailField(help_text="Lorem dolor")
    test_float = models.FloatField(help_text="Lorem dolor")
    test_bigint = models.BigIntegerField(help_text="Lorem dolor")
    test_positive_integer = models.PositiveIntegerField(help_text="Lorem dolor")
    test_decimal = models.DecimalField(max_digits=5, decimal_places=2, help_text="Lorem dolor")
    test_comma_separated_int = models.CharField(max_length=100, help_text='Lorem dolor', validators=[validate_comma_separated_integer_list])
    test_small_int = models.SmallIntegerField(help_text="Lorem dolor")
    test_nullbool = models.NullBooleanField(help_text="Lorem dolor")
    test_filepath = models.FilePathField(blank=True, help_text="Lorem dolor", path=".")
    test_positive_small_int = models.PositiveSmallIntegerField(help_text="Lorem dolor")

    def get_absolute_url(self):
        return ''

    class Meta:
        verbose_name = u'Test me'
        verbose_name_plural = u'Lot of Test me'


class TestMeProxyForFieldsets(TestMe):
    class Meta:
        proxy = True
        verbose_name = u'Test me fieldsets'
        verbose_name_plural = u'Lot of Test me fieldsets'


class TestThat(models.Model):
    that = models.ForeignKey(TestMe, help_text="Lorem dolor", on_delete=models.CASCADE)
    test_ip = models.GenericIPAddressField(help_text="Lorem dolor")
    test_url = models.URLField(help_text="Lorem dolor")
    test_int = models.IntegerField(help_text="Lorem dolor")
    test_date = models.DateField(help_text="Lorem dolor")
    test_bool = models.BooleanField(help_text="Lorem dolor", default=True)
    test_fk = models.ForeignKey('TestSortable', help_text="Lorem dolor", null=True, blank=True, on_delete=models.CASCADE)

    class Meta:
        verbose_name = u'Test that'
        verbose_name_plural = u'Lot of Test that'


class TestSortable(models.Model):
    that = models.ForeignKey(TestMe, on_delete=models.CASCADE)
    position = models.PositiveSmallIntegerField("Position")
    test_char = models.CharField(max_length=5)

    class Meta:
        ordering = ('position', )


test_django_admin_bootstrapped/test_django_admin_bootstrapped/settings.py
"""
Django settings for test_django_admin_bootstrapped project.

For more information on this file, see
https://docs.djangoproject.com/en/1.6/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.6/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
BASE_DIR = os.path.dirname(os.path.dirname(__file__))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.6/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '3!^@q06fn2-zl%2f%rmux58ybi9u=9k_lq^k*+^429foc#7fzn'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

TEMPLATE_DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    'bootstrap3',
    'django_admin_bootstrapped',
    'easy_thumbnails',
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    'test_django_admin_bootstrapped',
    'CapitalApp',
)

MIDDLEWARE = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

ROOT_URLCONF = 'test_django_admin_bootstrapped.urls'

WSGI_APPLICATION = 'test_django_admin_bootstrapped.wsgi.application'


# Database
# https://docs.djangoproject.com/en/1.6/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Internationalization
# https://docs.djangoproject.com/en/1.6/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

DAB_FIELD_RENDERER = 'django_admin_bootstrapped.renderers.BootstrapFieldRenderer'

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.6/howto/static-files/

STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'static')

EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'


test_django_admin_bootstrapped/test_django_admin_bootstrapped/migrations/0001_initial.py
# Generated by Django 2.2.6 on 2019-10-31 16:04

import django.core.validators
from django.db import migrations, models
import django.db.models.deletion
import re


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='TestMe',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('test_ip', models.GenericIPAddressField(help_text='Lorem dolor')),
                ('test_url', models.URLField(help_text='Lorem dolor')),
                ('test_int', models.IntegerField(help_text='Lorem dolor')),
                ('test_img', models.ImageField(blank=True, upload_to='dummy')),
                ('test_file', models.FileField(blank=True, upload_to='dummy')),
                ('test_date', models.DateField(help_text='Lorem dolor')),
                ('test_char', models.CharField(help_text='Lorem dolor', max_length=50)),
                ('test_bool', models.BooleanField(default=False, help_text='Lorem dolor')),
                ('test_time', models.TimeField(help_text='Lorem dolor')),
                ('test_slug', models.SlugField(help_text='Lorem dolor')),
                ('test_text', models.TextField(help_text='Lorem dolor')),
                ('test_email', models.EmailField(help_text='Lorem dolor', max_length=254)),
                ('test_float', models.FloatField(help_text='Lorem dolor')),
                ('test_bigint', models.BigIntegerField(help_text='Lorem dolor')),
                ('test_positive_integer', models.PositiveIntegerField(help_text='Lorem dolor')),
                ('test_decimal', models.DecimalField(decimal_places=2, help_text='Lorem dolor', max_digits=5)),
                ('test_comma_separated_int', models.CharField(help_text='Lorem dolor', max_length=100, validators=[django.core.validators.RegexValidator(re.compile('^\\d+(?:\\,\\d+)*\\Z'), code='invalid', message='Enter only digits separated by commas.')])),
                ('test_small_int', models.SmallIntegerField(help_text='Lorem dolor')),
                ('test_nullbool', models.NullBooleanField(help_text='Lorem dolor')),
                ('test_filepath', models.FilePathField(blank=True, help_text='Lorem dolor')),
                ('test_positive_small_int', models.PositiveSmallIntegerField(help_text='Lorem dolor')),
                ('test_m2m', models.ManyToManyField(blank=True, help_text='Lorem dolor', related_name='_testme_test_m2m_+', to='test_django_admin_bootstrapped.TestMe')),
            ],
            options={
                'verbose_name': 'Test me',
                'verbose_name_plural': 'Lot of Test me',
            },
        ),
        migrations.CreateModel(
            name='TestSortable',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('position', models.PositiveSmallIntegerField(verbose_name='Position')),
                ('test_char', models.CharField(max_length=5)),
                ('that', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='test_django_admin_bootstrapped.TestMe')),
            ],
            options={
                'ordering': ('position',),
            },
        ),
        migrations.CreateModel(
            name='TestThat',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('test_ip', models.GenericIPAddressField(help_text='Lorem dolor')),
                ('test_url', models.URLField(help_text='Lorem dolor')),
                ('test_int', models.IntegerField(help_text='Lorem dolor')),
                ('test_date', models.DateField(help_text='Lorem dolor')),
                ('test_bool', models.BooleanField(default=True, help_text='Lorem dolor')),
                ('test_fk', models.ForeignKey(blank=True, help_text='Lorem dolor', null=True, on_delete=django.db.models.deletion.CASCADE, to='test_django_admin_bootstrapped.TestSortable')),
                ('that', models.ForeignKey(help_text='Lorem dolor', on_delete=django.db.models.deletion.CASCADE, to='test_django_admin_bootstrapped.TestMe')),
            ],
            options={
                'verbose_name': 'Test that',
                'verbose_name_plural': 'Lot of Test that',
            },
        ),
        migrations.CreateModel(
            name='TestMeProxyForFieldsets',
            fields=[
            ],
            options={
                'verbose_name': 'Test me fieldsets',
                'verbose_name_plural': 'Lot of Test me fieldsets',
                'proxy': True,
                'indexes': [],
                'constraints': [],
            },
            bases=('test_django_admin_bootstrapped.testme',),
        ),
    ]
